File: vsuetapi/core/src/main/java/ru/vsuetapp/controller/AdminController.java
package ru.vsuetapp.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import ru.vsuetapp.dto.timetableJSON.TimetableDto;
import ru.vsuetapp.model.Faculty;
import ru.vsuetapp.model.GroupTimetable;
import ru.vsuetapp.model.TeacherTimetable;
import ru.vsuetapp.model.User;
import ru.vsuetapp.service.AdminService;

import java.util.Map;

@RestController
@RequestMapping("/api/admin")
@RequiredArgsConstructor
public class AdminController {

    private final AdminService adminService;

    // =============== FACULTY ===============
    @PostMapping("/faculty")
    public ResponseEntity<?> createDeanUser(
            @RequestParam String facultyName
    ) {
        try {
            Faculty faculty = adminService.createFaculty(facultyName);
            return ResponseEntity.ok(faculty);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @DeleteMapping("/faculty/{id}")
    public ResponseEntity<?> deleteFaculty(@PathVariable Long id) {
        try {
            adminService.deleteFacultyById(id);
            return ResponseEntity.ok(Map.of("message", "Факультет успешно удалён"));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    // =============== CREATE USERS ===============
    @PostMapping("/users/dean")
    public ResponseEntity<?> createDeanUser(
            @RequestParam String username,
            @RequestParam String password,
            @RequestParam String deanName,
            @RequestParam Long facultyId
    ) {
        try {
            User user = adminService.createDeanUser(username, password, deanName, facultyId);
            return ResponseEntity.ok(user);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/users/student")
    public ResponseEntity<?> createStudentUser(
            @RequestParam String username,
            @RequestParam String password,
            @RequestParam String studentName,
            @RequestParam Long groupId,
            @RequestParam String zachNumber
    ) {
        try {
            User user = adminService.createStudentUser(username, password, studentName, groupId, zachNumber);
            return ResponseEntity.ok(user);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/users/teacher")
    public ResponseEntity<?> createTeacherUser(
            @RequestParam String username,
            @RequestParam String password,
            @RequestParam String teacherName
    ) {
        try {
            User user = adminService.createTeacherUser(username, password, teacherName);
            return ResponseEntity.ok(user);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/users/admin")
    public ResponseEntity<?> createAdminUser(
            @RequestParam String username,
            @RequestParam String password
    ) {
        try {
            User user = adminService.createAdminUser(username, password);
            return ResponseEntity.ok(user);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }


    // =============== DELETE USERS ===============
    @DeleteMapping("/users/dean/{userId}")
    public ResponseEntity<String> deleteDeanUser(@PathVariable Long userId) {
        adminService.deleteDeanUser(userId);
        return ResponseEntity.ok("Декан и его информация успешно удалены");
    }

    @DeleteMapping("/users/student/{userId}")
    public ResponseEntity<String> deleteStudentUser(@PathVariable Long userId) {
        adminService.deleteStudentUser(userId);
        return ResponseEntity.ok("Студент и его информация успешно удалены");
    }

    @DeleteMapping("/users/teacher/{userId}")
    public ResponseEntity<String> deleteTeacherUser(@PathVariable Long userId) {
        adminService.deleteTeacherUser(userId);
        return ResponseEntity.ok("Преподаватель и его информация успешно удалены");
    }


    // =============== TEACHER TIMETABLE ===============
    @PostMapping("/timetable/teacher/{teacherId}")
    public ResponseEntity<TeacherTimetable> createTeacherTimetable(
            @PathVariable Long teacherId,
            @RequestBody TimetableDto timetableDto
    ) {
        return ResponseEntity.ok(adminService.createTeacherTimetable(teacherId, timetableDto));
    }

    @DeleteMapping("/timetable/teacher/{id}")
    public ResponseEntity<String> deleteTeacherTimetable(@PathVariable Long id) {
        adminService.deleteTeacherTimetable(id);
        return ResponseEntity.ok("Расписание преподавателя удалено");
    }

    // =============== GROUP TIMETABLE ===============
    @PostMapping("/timetable/groups/{groupId}")
    public ResponseEntity<GroupTimetable> createStudentTimetable(
            @PathVariable Long groupId,
            @RequestBody TimetableDto timetableDto
    ) {
        GroupTimetable timetable = adminService.createStudentTimetable(groupId, timetableDto);
        return ResponseEntity.ok(timetable);
    }

    @DeleteMapping("/timetable/groups/{groupId}")
    public ResponseEntity<String> deleteStudentTimetable(@PathVariable Long groupId) {
        adminService.deleteTimetableByGroupId(groupId);
        return ResponseEntity.ok("Расписание группы удалено");
    }
}

File: fastapi/app/controllers/admin_controller.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from app.config.database import get_db
from app.models.pydantic_models import (
    FacultyCreate, FacultyResponse, UserResponse, TimetableDto,
    GroupTimetable, TeacherTimetable
)
from app.services.admin_service import AdminService

admin_router = APIRouter(prefix="/api/admin", tags=["admin"])


@admin_router.post("/faculty", response_model=FacultyResponse)
async def create_faculty(
    faculty_name: str = Query(..., description="Faculty name"),
    db: Session = Depends(get_db)
):
    """Create a new faculty"""
    try:
        admin_service = AdminService(db)
        faculty = admin_service.create_faculty(faculty_name)
        return faculty
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/faculty/{faculty_id}")
async def delete_faculty(faculty_id: int, db: Session = Depends(get_db)):
    """Delete faculty by ID"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_faculty_by_id(faculty_id)
        return {"message": "Факультет успешно удалён"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/users/dean", response_model=UserResponse)
async def create_dean_user(
    username: str = Query(..., description="Username"),
    password: str = Query(..., description="Password"),
    dean_name: str = Query(..., description="Dean name"),
    faculty_id: int = Query(..., description="Faculty ID"),
    db: Session = Depends(get_db)
):
    """Create a new dean user"""
    try:
        admin_service = AdminService(db)
        user = admin_service.create_dean_user(username, password, dean_name, faculty_id)
        return user
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/users/student", response_model=UserResponse)
async def create_student_user(
    username: str = Query(..., description="Username"),
    password: str = Query(..., description="Password"),
    student_name: str = Query(..., description="Student name"),
    group_id: int = Query(..., description="Group ID"),
    zach_number: str = Query(..., description="Zach number"),
    db: Session = Depends(get_db)
):
    """Create a new student user"""
    try:
        admin_service = AdminService(db)
        user = admin_service.create_student_user(username, password, student_name, group_id, zach_number)
        return user
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/users/teacher", response_model=UserResponse)
async def create_teacher_user(
    username: str = Query(..., description="Username"),
    password: str = Query(..., description="Password"),
    teacher_name: str = Query(..., description="Teacher name"),
    db: Session = Depends(get_db)
):
    """Create a new teacher user"""
    try:
        admin_service = AdminService(db)
        user = admin_service.create_teacher_user(username, password, teacher_name)
        return user
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/users/admin", response_model=UserResponse)
async def create_admin_user(
    username: str = Query(..., description="Username"),
    password: str = Query(..., description="Password"),
    db: Session = Depends(get_db)
):
    """Create a new admin user"""
    try:
        admin_service = AdminService(db)
        user = admin_service.create_admin_user(username, password)
        return user
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/users/dean/{user_id}")
async def delete_dean_user(user_id: int, db: Session = Depends(get_db)):
    """Delete dean user"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_dean_user(user_id)
        return {"message": "Декан и его информация успешно удалены"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/users/student/{user_id}")
async def delete_student_user(user_id: int, db: Session = Depends(get_db)):
    """Delete student user"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_student_user(user_id)
        return {"message": "Студент и его информация успешно удалены"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/users/teacher/{user_id}")
async def delete_teacher_user(user_id: int, db: Session = Depends(get_db)):
    """Delete teacher user"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_teacher_user(user_id)
        return {"message": "Преподаватель и его информация успешно удалены"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/timetable/teacher/{teacher_id}", response_model=TeacherTimetable)
async def create_teacher_timetable(
    teacher_id: int,
    timetable_dto: TimetableDto,
    db: Session = Depends(get_db)
):
    """Create teacher timetable"""
    try:
        admin_service = AdminService(db)
        timetable = admin_service.create_teacher_timetable(teacher_id, timetable_dto)
        return timetable
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/timetable/teacher/{teacher_id}")
async def delete_teacher_timetable(teacher_id: int, db: Session = Depends(get_db)):
    """Delete teacher timetable"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_teacher_timetable(teacher_id)
        return {"message": "Расписание преподавателя удалено"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.post("/timetable/groups/{group_id}", response_model=GroupTimetable)
async def create_student_timetable(
    group_id: int,
    timetable_dto: TimetableDto,
    db: Session = Depends(get_db)
):
    """Create group timetable"""
    try:
        admin_service = AdminService(db)
        timetable = admin_service.create_student_timetable(group_id, timetable_dto)
        return timetable
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@admin_router.delete("/timetable/groups/{group_id}")
async def delete_student_timetable(group_id: int, db: Session = Depends(get_db)):
    """Delete group timetable"""
    try:
        admin_service = AdminService(db)
        admin_service.delete_timetable_by_group_id(group_id)
        return {"message": "Расписание группы удалено"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})
